docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,388 @@
+often more important to have guaranteed conversions, even if copies are created.
+
+Basic interoperability
+----------------------
+
+NumPy arrays have no conception of the GPU, and so all conversions require a
+``cpu`` tensor. Similarly, the **computational graph** cannot be represented in
+```numpy``, and so the pre-requisite to have an operation or object without one,
+i.e. only leaf nodes are supported.
+
+.. doctest::
+
+   >>> b_torch = torch.ones(3)
+   >>> assert(b_torch.requires_grad == False)
+   >>> assert(b_torch.is_cuda == False)
+   >>> b_torch.numpy()
+   array([1., 1., 1.], dtype=float32)
+
+PyTorch will raise exceptions instead of allowing operations that may lead to
+incorrect results. As the conversion to a NumPy array may potentially be a lossy
+operation (the computational graph), there are no inbuilt operations to do
+something which unconditionally returns a NumPy array. Practically speaking, a
+function can be created locally to return a NumPy array, for say, visualizations.
+
+.. code-block:: python
+
+   # Possibly lossy function for visualizations
+   def uncond_numpy(torch_tensor):
+        if torch_tensor.is_cuda:
+           torch_tensor = torch_tensor.cpu()
+        if torch_tensor.requires_grad == False:
+            return torch_tensor.numpy()
+        else:
+            return torch_tensor.detach().numpy()
+
+Operations
+----------
+
+All ``torch`` operators will helpfully fail with a ``TypeError`` if called with
+``numpy`` arrays. However, for **numpy functions**, using a ``torch.Tensor`` with
+an ``np.ndarray`` will return a ``torch.Tensor`` without creating any additional
+copies.
+
+ - Due to the :meth:`torch.Tensor.__array__()` implementation, a
+   ``np.ndarray`` which shares memory with the ``torch.Tensor`` is used for the
+   operation.
+ - The return type functionality is defined by
+   :meth:`torch.Tensor.__array_wrap__()`, and calls ``torch.from_numpy()``
+   internally.
+
+As a concrete example, consider the following snippet:
+
+.. doctest::
+
+   >>> a_np = np.ones(3)
+   >>> a_np.dtype
+   dtype('float64')
+   >>> b_torch = torch.ones(3)
+   >>> b_torch.dtype
+   torch.float32
+   >>> torch.add(a_np, b_torch)
+   Traceback (most recent call last):
+   ...
+   TypeError: add(): argument 'input' (position 1) must be Tensor, not numpy.ndarray
+   >>> b_torch + a_np
+   tensor([2., 2., 2.], dtype=torch.float64)
+   >>> a_np + b_torch
+   Traceback (most recent call last):
+   ...
+   TypeError: Concatenation operation is not implemented for NumPy arrays, use np.concatenate() instead. Please do not rely on this error; it may not be given on all Python implementations.
+   >>> np.add(a_np, b_torch)
+   tensor([2., 2., 2.], dtype=torch.float64)
+
+To recognize the root cause of the errors and the legitimacy of the results above, we will recall that:
+
+- The `Python data model`_ specifies that the ``__radd__`` function is to be
```

Comment:
I'm not sure this detail about these particular operators is useful. 

In general NumPy arrays can't be used in torch operations and SOME NumPy operations will work with torch tensors but they don't work on them in general.

Reply:
Abstracted into an optional dropdown.